### PR TITLE
MVP JavaScript Console

### PR DESCRIPTION
**Pull request by outoftime**
Introduces a bare-bones JavaScript console, not ready for general release but enough to start playing with in experimental mode (`https://popcode.org?experimental`) and iterating on.

Specifically, support the following workflow:

* Type an expression into the console, press enter
* Expression is evaluated in the current preview page environment
* Result of evaluation is printed in the console

The is the first step in the project described by #312; see that issue description for a full list of steps (and to get a better sense of what this PR is *not* yet trying to do).

Here is the console in action:

console

### CODE REVIEW

**Comment by jwang1919** [comment]
Great stuff @outoftime!

Can't comment much about the code because I honestly don't understand it well enough to form an expert opinion on it, but I can comment on the functionality.

At it's current incarnation, I think it is almost good enough to be usable in class. I think we can start trying to integrate it into some lessons after a clear button/command gets implemented.

Also found some strange bugs when attempting to type in an object.
Typing `{` and enter, causes Popcode to freeze
Typing `{}` and enter, doesn't return any feedback
Typing `}` and enter, returns expected result.

File: src/util/generatePreview.js [hunk 1]
```diff
@@ -80,6 +118,7 @@ export class PreviewGenerator {
     this._addCss();
     if (options.propagateErrorsToParent) {
       this._addErrorHandling();
+      this._addMessageHandling();
```
Comment (outoftime) [inline src/util/generatePreview.js]: Should be its own option

File: src/components/Preview.jsx [hunk 2]
```diff
@@ -15,10 +19,17 @@ export default function Preview({
...
   }
 
-  const projectFrames = compiledProjects.map(({source, timestamp}) => (
+  const projectFrames = compiledProjects.map(({source, timestamp}, key) => (
     <PreviewFrame
+      consoleEntries={
+        key === compiledProjects.keySeq().last() ?
+          consoleEntries :
+          new OrderedMap()
+      }
```
Comment (outoftime) [inline src/components/Preview.jsx]: Is there a less clunky way to do this?

File: src/components/ConsoleEntry.jsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {ConsoleEntry as ConsoleEntryRecord} from '../records';
+
+export default function ConsoleEntry({
+  entry: {expression, status, value, error},
+}) {
+  let output;
+
+  if (status === 'evaluated' && value !== undefined) {
```
Comment (outoftime) [inline src/components/ConsoleEntry.jsx]: Do we need `status`?

File: src/components/ConsoleEntry.jsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {ConsoleEntry as ConsoleEntryRecord} from '../records';
+
+export default function ConsoleEntry({
+  entry: {expression, status, value, error},
+}) {
+  let output;
```
Comment (outoftime) [inline src/components/ConsoleEntry.jsx]: `ConsoleOutput` component

File: src/components/Console.jsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  return session;
+}
+
+export default class Console extends Component {
+  constructor() {
+    super();
+    bindAll(this, '_handleInput');
+  }
+
+  _handleInput(containerElement) {
```
Comment (outoftime) [inline src/components/Console.jsx]: Decompose into a `<ConsoleInput>` component

**Pull request merged**